src/gt4py/next/ffront/nd_array_field.py

```diff
@@ -0,0 +1,245 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import annotations
+
+import abc
+import dataclasses
+import functools
+import typing
+from collections.abc import Callable
+from types import ModuleType
+from typing import Any, ClassVar, Optional, ParamSpec, TypeVar
+
+import numpy as np
+from numpy import typing as npt
+
+from gt4py.next import common
+
+
+try:
+    import cupy as cp
+except ImportError:
+    cp: Optional[ModuleType] = None  # type:ignore[no-redef]
```

Comment:
played with different things to make mypy happy, in the end the only thing I found was introduce a `HAS_CP` flag instead of dispatching on `cp` directly.